### PR TITLE
Support tuples as a result and input type

### PR DESCRIPTION
**Pull request by savolgin**
Added support for operations results as tuples. Issue #6

### CODE REVIEW

**Review by olegrok — COMMENTED** [review]
Thanks for your patch!

Do you understand that `crud.init` is executed on _storage_ but all flattening/unflattening takes place on the router?

It makes you approach a bit wrong. Also I don't like "tuples_as_map" name. IMO it should be something like "use_tomap=true/false" (maybe @dokshina will purpose a better name for it).

Next, seems the correct way to solve your task - it save "call" result to "buffer" and then convert to formatted tuple (since 2.2 format is stored in conn.space.<space_name>._cdata_format variable). I'm not sure it's easy to implement (https://github.com/tarantool/tarantool/blob/0018398db632b3d0131ca8dcc1f65315039bbf98/src/box/lua/net_box.lua#L73) and some internal magic is required. But after that user will get formatted tuples that will look like simple tuples from spaces.

**Comment by dokshina** [comment]
I agree with @olegrok.

Since your changes affects current API, I would like you to add RFC to the issue and describe new flags and configuring router. We need to discuss this changes. I sure that it's the right way to make `CRUD` pretty and useful.

1. About the new flag name, I suggest `tomap` or `use_tomap`, or maybe `tuples_tomap`. I will cal it `tomap` in this comment.

2. Yes, `crud.init` is called on storage which operates with tuples and know nothing about router's staff such as current select state or tuples flattening/unflattenning. There is no reason to pass  `tomap` parameter to `crud.init`. I like adding a new flag to each call, and I agree that it will be useful to set it once. Maybe we need something like `crud.cfg` or `crud.router.cfg` to specify such "global" options on router side.

3. I would be great to return formatted tuples, but it can be implemented later in a separated patch.

**Comment by knazarov** [comment]
I'd argue that the key problem here is with the connectors. When you use connectors to access regular non-sharded Tarantool, you get tuples that have "format", because the server provides you with it.
If the router doesn't have the format for those tuples, it will strip the format, and you'd receive tuples on the connector side that don't have information about field names. 

This is a major usability problem and (i suppose) the reason why current implementation uses maps. 

I understand that maps perform worse than tuples, and I agree that we should support tuples as well (maybe even by default), but we should somehow solve the problem with the format. Which brings us to the https://github.com/tarantool/ddl module. I expect that for this to be possible, we need to have the same spaces created on routers.

**Comment by akudiyar** [comment]
Not all connectors support schema loading, and not all spaces have the schema defined. So working with standard tuples is more basic than working with tuples with fields.

Although, I agree with @knazarov about the lack of schema on routers, and we overcome this by providing recommendations of using DDL module for retrieving schemas or writing custom API functions when using the new Java connector. The "tomap" conversion is therefore performed inside the connector, which allows working seamlessly with any setup.

**Comment by akudiyar** [comment]
@dokshina The short proposal is described in #6 . I suggest renaming the option and so not set any options in init -- let it always be an explicit option for each request.

Is this enough?

**Review by akudiyar — CHANGES_REQUESTED** [review]

File: README.md [hunk 1]
```diff
@@ -7,11 +7,16 @@ It also provides the `crud-storage` role for
 ## API
 
 The CRUD operations should be called from router.
-All storage replica sets should call `crud.init()`
+All storage replica sets should call `crud.init(opts)`
```
[the remark this answers is not in the document]
Comment (savolgin) [reply README.md]: Global opts removed

File: crud/common/utils.lua [hunk 2]
```diff
@@ -55,9 +64,21 @@ function utils.flatten(object, space_format)
...
+        if system_fields[field_format.name] then
+            return fieldno
+        end
+    end
+    return nil, UnflattenError:new("Field `bucket_id` not found in space format")
+end
 
+function utils.unflatten(tuple, space_format, as_map)
+    if tuple == nil then return nil end
+    local result_as_map = as_map
```
[the remark this answers is not in the document]
Comment (savolgin) [reply crud/common/utils.lua]: Parameter removed

File: crud/common/utils.lua [hunk 3]
```diff
@@ -70,7 +91,11 @@ function utils.unflatten(tuple, space_format)
         object[field_format.name] = value
     end
 
-    return object
+    if result_as_map then
```
[the remark this answers is not in the document]
Comment (savolgin) [reply crud/common/utils.lua]: fixed: unflatten not calling for if tuples_tomap == false

File: crud/delete.lua [hunk 4]
```diff
@@ -12,6 +12,8 @@ local delete = {}
 
 local DELETE_FUNC_NAME = '__delete'
 
+local format_tuples_as_map = true
```
[the remark this answers is not in the document]
Comment (savolgin) [reply crud/delete.lua]: removed

**Review by akudiyar — APPROVED** [review]

**Comment by dokshina** [comment]
I think, we can close this PR.

It was implemented in #39 (merged) and #40, #42, #43 PRs that will be merged soon.

**Pull request closed without merging**